src/compiler/emitter.ts

```diff
@@ -1179,4394 +1178,5531 @@ namespace ts {
+                        break;
+                    }
+                    case BinaryExpressionState.Right: {
+                        const isCommaOperator = node.operatorToken.kind !== SyntaxKind.CommaToken;
+                        const linesBeforeOperator = getLinesBetweenNodes(node, node.left, node.operatorToken);
+                        const linesAfterOperator = getLinesBetweenNodes(node, node.operatorToken, node.right);
+                        writeLinesAndIndent(linesBeforeOperator, isCommaOperator);
+                        emitLeadingCommentsOfPosition(node.operatorToken.pos);
+                        writeTokenNode(node.operatorToken, node.operatorToken.kind === SyntaxKind.InKeyword ? writeKeyword : writeOperator);
+                        emitTrailingCommentsOfPosition(node.operatorToken.end, /*prefixSpace*/ true); // Binary operators should have a space before the comment starts
+                        writeLinesAndIndent(linesAfterOperator, /*writeSpaceIfNotIndenting*/ true);
+                        maybePipelineEmitExpression(node.right);
+                        break;
+                    }
+                    case BinaryExpressionState.Finish: {
+                        const linesBeforeOperator = getLinesBetweenNodes(node, node.left, node.operatorToken);
+                        const linesAfterOperator = getLinesBetweenNodes(node, node.operatorToken, node.right);
+                        decreaseIndentIf(linesBeforeOperator, linesAfterOperator);
+                        stackIndex--;
+                        break;
+                    }
+                    default: return Debug.fail(`Invalid state ${stateStack[stackIndex]} for emitBinaryExpressionWorker`);
                 }
-                writeSpace();
             }
-            writePunctuation("[");
 
-            pipelineEmit(EmitHint.MappedTypeParameter, node.typeParameter);
-            if (node.nameType) {
-                writeSpace();
-                writeKeyword("as");
-                writeSpace();
-                emit(node.nameType);
-            }
+            function maybePipelineEmitExpression(next: Expression) {
+                // Advance the state of this unit of work,
+                stateStack[stackIndex]++;
 
-            writePunctuation("]");
-            if (node.questionToken) {
-                emit(node.questionToken);
-                if (node.questionToken.kind !== SyntaxKind.QuestionToken) {
-                    writePunctuation("?");
+                // Then actually do the work of emitting the node `next` returned by the prior state
+
+                // The following section should be identical to `pipelineEmit` save it assumes EmitHint.Expression and offloads
+                // binary expression handling, where possible, to the contained work queue
+
+                // #region trampolinePipelineEmit
+                if (!shouldEmitComments(next) && !shouldEmitSourceMaps(next) && isBinaryExpression(next)) {

```

Comment:
Actually, now that we no longer use the `pipelinePhase` system as heavily, is it easier to decorate the sourcemap/comment emit calls inside the trampoline (if needed)? Or are those still too hard to call on individual nodes without recurring? Our inability to trampoline when those are on is more of an unfortunate implementation detail, than a goal - I'd love if we could clean it up.

Reply:
We probably could, but there is a fair amount of state that needs to be tracked (or recomputed) as you walk the tree.